### ISSUE TITLE
[c23] support `unsequenced/reproducible` attributes

### ISSUE THREAD

**Issue by SchrodingerZhu** [llvm/llvm-project#84406]
We probably want to introduce `unsequenced/reproducible` attributes to enhance C23 compatibility. This requires changes both in clang and in the libc. These attributes are similar to `const/pure` but does have subtle differences.

See 6.7.12.7 in https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf.

At libc's side, most bitwise operations would be annotated with unsequenced to allow further optimizations. Although many optimization opportunities should already be captured libcalls.

**Comment by SchrodingerZhu** [comment]
libc's change is also mentioned in https://github.com/llvm/llvm-project/issues/83165

**Comment by shafik** [comment]
CC @AaronBallman @erichkeane

**Comment by AaronBallman** [comment]
Yup, we don't yet implement this feature (https://clang.llvm.org/c_status.html tracks the status of any C standard feature, so we don't usually need an issue reported for missing standard features, same as in C++).

**Comment by p7r0x7** [comment]
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2956.htm clearly states __attribute__((const)) and __attribute__((pure)) are valid (but imperfect/unideal) implementations of [[unsequenced]] and [[reproducible]], respectively; I strongly recommend applying __attribute__((const)) and __attribute__((pure)) optimization codepaths to uses of [[unsequenced]] and [[reproducible]]. I cannot understand why this has not yet been done, even as a stopgap for making the optimization paths more correct and ideal.

**Comment by SamB** [comment]
The GCC patch might be interesting: https://inbox.sourceware.org/gcc-patches/ZqkdBSetdAYlWU7x@tucnak/t/
